runtime/health.go

```diff
@@ -57,6 +52,163 @@ func (r *Runtime) Health(ctx context.Context) (*Health, error) {
+	res := &InstanceHealth{}
+	// check repo error
+	repo, rr, err := r.Repo(ctx, instanceID)
+	if err != nil {
+		res.Repo = err.Error()
+	} else {
+		err = repo.(drivers.Handle).Ping(ctx)
+		if err != nil {
+			res.Repo = err.Error()
+		}
+		rr()
+	}
+
+	ctrl, err := r.Controller(ctx, instanceID)
+	if err != nil {
+		res.Controller = err.Error()
+		return res, nil
+	}
+
+	// check olap error
+	olap, release, err := r.OLAP(ctx, ctrl.InstanceID, "")
+	if err != nil {
+		res.OLAP = err.Error()
+		return res, nil
+	}
+	defer release()
+
+	resources, err := ctrl.List(ctx, ResourceKindMetricsView, "", false)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheEnabled := healthCheckCacheEnabled(olap)
+	if cacheEnabled {
+		instanceHealth, ok := r.cachedInstanceHealth(ctx, ctrl.InstanceID, ctrl.catalog.version, resources)
+		if ok {
+			res.OLAP = instanceHealth.OLAP
+			res.Dashboards = instanceHealth.Dashboards
+			return res, nil
+		}
+	}
+
+	err = olap.(drivers.Handle).Ping(ctx)
+	if err != nil {
+		res.OLAP = err.Error()
+	} else if query != nil {
+		// run queries against dashboards
+		res.Dashboards = make(map[string]string, len(resources))
+		for _, mv := range resources {
+			q, err := query(ctx, ctrl.InstanceID, mv.Meta.Name.Name)
+			if err != nil {
+				res.Dashboards[mv.Meta.Name.Name] = err.Error()
+				continue
+			}
+			err = r.Query(ctx, ctrl.InstanceID, q, 1)
+			if err != nil {
+				res.Dashboards[mv.Meta.Name.Name] = err.Error()
+			}
+		}
+	}
+	if !cacheEnabled {
+		return res, nil
+	}
+
+	// save to cache
+	hash, err := healthResultHash(ctrl.catalog.version, resources)
+	if err != nil {
+		return nil, err
+	}
+	res.Hash = hash
+
+	bytes, err := json.Marshal(res)
+	if err != nil {
+		return nil, err
+	}
+
+	catalog, release, err := r.Catalog(ctx, instanceID)
+	if err != nil {
+		return nil, err
+	}
+	defer release()
+
+	err = catalog.UpsertInstanceHealth(ctx, &drivers.InstanceHealth{
+		InstanceID: instanceID,
+		Health:     bytes,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (r *Runtime) cachedInstanceHealth(ctx context.Context, instanceID string, ctrlVersion int64, resources []*runtimev1.Resource) (*InstanceHealth, bool) {
+	catalog, release, err := r.Catalog(ctx, instanceID)
+	if err != nil {
+		return nil, false
+	}
+	defer release()
+
+	cached, err := catalog.FindInstanceHealth(ctx, instanceID)
+	if err != nil {
+		return nil, false
+	}
+
+	c := &InstanceHealth{}
+	err = json.Unmarshal(cached.Health, c)
+	if err != nil {
+		return nil, false
+	}
+
+	hash, err := healthResultHash(ctrlVersion, resources)
+	if err != nil || hash != c.Hash {
+		return nil, false
+	}
+	return c, true
+}
+
+func healthCheckCacheEnabled(olap drivers.OLAPStore) bool {
+	if olap.Dialect() != drivers.DialectClickHouse {
+		return false
+	}
+	m := olap.(drivers.Handle).Config()
+	enabled, ok := m["enable_health_check_cache"].(bool)
+	if !ok {
+		return true
+	}
+	return enabled
+}
```

Comment:
Accessing a config key like this feels a bit hacky – how do you feel about adding something like a `MaybeScaledToZero() bool` on the OLAP driver that internally checks a config key in ClickHouse (which in the future could be powered by actual API calls to CH cloud)?

Reply:
Actually `MaybeScaledToZero` can improve this logic but we would still need to check this config. A use case could be to wake up the service even if it is scaled to zero for health checks (may be for debugging).
If it feels hacky to access configs outside of clickhouse or olap then a method `HealthCheckCacheEnabled` can be added on the OLAP interface.

Overall I think we should take a call about the `Config` method. May be it should be removed and connections can expose methods to access their configs directly like `Config()` on s3 connection which returns `S3Config` struct.

Reply:
Agree removing the `Config` method makes sense, until then let's avoid using it